### ISSUE TITLE
Transpilation generates unreachable code

### ISSUE THREAD

**Issue by joscha** [Microsoft/TypeScript#23732]
**TypeScript Version:**  2.8.1 and Playground

**Search Terms:** `unreachable code`, `rest spread`

**Code**

```ts
const bla = () => ({ x: 1, y: {} });

function z() {
    let y: any;
    let x: number;
    ({ x, ...y } = bla());
    return { x, y };
}
```

transpiles to:
```js
// ... removed some spread boilerplate here
var bla = function () { return ({ x: 1, y: {} }); };
function z() {
    var y;
    var x;
    (_a = bla(), (x = _a.x, _a), y = __rest(_a, ["x"]));
    return { x: x, y: y };
    var _a;
}
```

which has a `var _a;` that is unreachable and Closure Compiler warns about.

**Expected behavior:**
`var _a` to be in the block with the other variable declarations at the beginning of the function.

**Actual behavior:**
Transpilation produces unreachable code.

**Playground Link:** https://www.typescriptlang.org/play/index.html#src=const%20bla%20%3D%20()%20%3D%3E%20(%7B%20x%3A%201%2C%20y%3A%20%7B%7D%20%7D)%3B%0D%0A%0D%0Afunction%20z()%20%7B%0D%0A%20%20%20%20let%20y%3A%20any%3B%0D%0A%20%20%20%20let%20x%3A%20number%3B%0D%0A%20%20%20%20(%7B%20x%2C%20...y%20%7D%20%3D%20bla())%3B%0D%0A%20%20%20%20return%20%7B%20x%2C%20y%20%7D%3B%0D%0A%7D%0D%0A

**Related Issues:**

**Comment by Kovensky** [comment]
This is Closure Compiler being silly; `var _a;` is always reachable because any `var` declaration (without assignment) is always reachable.

**Comment by joscha** [comment]
> This is Closure Compiler being silly; var _a; is always reachable because any var declaration (without assignment) is always reachable, due to hoisting.

agreed that the generated code is executing as expected, but is there a reason we don't put the var at the top?

**Comment by j-oliveras** [comment]
Most likely because the emitter don't know the temporal variables needed after the end of emit and can't put the declarations at start because is already emitted.

I don't know if this is the case, I only guess.

**Comment by mhegazy** [comment]
There are historical reasons why the var is at the end of the scope. the old emitter was one pass emitter, so it only knew about the vars when the function transplilation was complete. the new emitter is a transformation based one, so we do have the list before we start generating the output of the function.

we would be happy to take a PR to emit the vars at the top of the scope.

**Comment by mhegazy** [comment]
someone needs to file an issue for closure nevertheless. `var _a` is 1. not code and 2. not unreachable.

**Comment by mhegazy** [comment]
Should be fixed by https://github.com/Microsoft/TypeScript/pull/23956
